### ISSUE TITLE
Various issues

### ISSUE THREAD

**Issue by vanhauser-thc** [junxzm1990/afl-pt#2]
I installed docker image ubuntu:14.04 and a lot of necessary dependencies.

the build of binutils fails:

```
./afl-pt/afl-2.42b/test_progs/binutils-2.29/build
[...]
/bin/bash ../../binutils/../ylwrap ../../binutils/syslex.l .c syslex.c -- ./afl-pt/afl-2.42b/test_progs/binutils-2.29/missing flex  
lex.yysyslex.c is unchanged
gcc -c -I. -I../../binutils -W -Wall -Wstrict-prototypes -Wmissing-prototypes -Wshadow -I../../binutils/../zlib -g -O2  ../../binutils/syslex_wrap.c
../../binutils/syslex_wrap.c:25:20: fatal error: syslex.c: No such file or directory
 #include "syslex.c"
                    ^
compilation terminated.
make[2]: *** [syslex_wrap.o] Error 1
```
but as this is only for the example I used something else instead, libtiff.

when libtiff is compiled it works fine. then I patch it with ./patch-bin.sh.
Afterwards it can not be run anymore:
```
lt-thumbnail: error while loading shared libraries: libm.so.6: cannot open shared object file: No such file or directory
```
then setting *LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/lib:/lib/x86_64-linux-gnu* lets it execute. However it does not run without any output

```
root@9f2aa2113a36:/prg/tmp/afl-pt/afl-2.42b# ldd /prg/tmp/afl-pt/tiff-4.0.4/tools/.libs/lt-thumbnail
	linux-vdso.so.1 =>  (0x00007ffff7ffa000)
	libtiff.so.5 => /prg/tmp/afl-pt/tiff-4.0.4/libtiff/.libs/libtiff.so.5 (0x00007ffff7b78000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ffff7872000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ffff74a9000)
	/prg/tmp/afl-pt/afl-2.42b/pt_mode/glibc-2.19/build/elf/ld.so => /lib64/ld-linux-x86-64.so.2 (0x00007ffff7dda000)
```

An strace shows this:
```
23    open("/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
23    read(3, "\177ELF...................
23    fstat(3, {st_dev=makedev(0, 45), st_ino=22020230, st_mode=S_IFREG|0755, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=3632, st_size=1857312, st_atime=2019/03/27-12:49:24, st_mtime=2019/03/27-12:49:24, st_ctime=2019/06/04-09:46:33}) = 0
23    mmap(NULL, 3965632, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7ffff74ab000
23    mprotect(0x7ffff7669000, 2097152, PROT_NONE) = 0
23    mmap(0x7ffff7869000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1be000) = 0x7ffff7869000
23    mmap(0x7ffff786f000, 17088, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7ffff786f000
23    close(3)                          = 0
23    mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ffff7ff4000
23    mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ffff7ff2000
23    arch_prctl(ARCH_SET_FS, 0x7ffff7ff2740) = 0
23    mprotect(0x7ffff7869000, 16384, PROT_READ) = 0
23    mprotect(0x7ffff7b78000, 4096, PROT_READ) = 0
23    mprotect(0x7ffff7dd7000, 4096, PROT_READ) = 0
23    mprotect(0x602000, 4096, PROT_READ) = 0
23    mprotect(0x7ffff7ffc000, 4096, PROT_READ) = 0
23    write(196, "\0\0\0\0", 4)         = -1 EBADF (Bad file descriptor)
23    exit_group(1)                     = ?
23    +++ exited with 1 +++
```

so then I thought maybe fd 196 is used to communicate to AFL-PT so I tried to run it with afl-fuzz:
```
afl-2.42b# ./pt-fuzz-fast -P -i in -o out -- /prg/tmp/afl-pt/tiff-4.0.4/tools/.libs/lt-thumbnail @@ /dev/null
afl-fuzz 2.42b by <lcamtuf@google.com>
[+] You have 8 CPU cores and 1 runnable tasks (utilization: 12%).
[+] Try parallel jobs - see /usr/local/share/doc/afl/parallel_fuzzing.txt.
[*] Checking CPU core loadout...
[+] Found a free CPU core, binding to #0.
[*] Checking core_pattern...
[*] Checking CPU scaling governor...
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Scanning 'in'...
[+] No auto-generated dictionary tokens to reuse.
[*] Creating hard links for all input files...
[*] Validating target binary...
[*] Attempting dry run with 'id:000000,orig:input-0.tiff'...
[*] Spinning up the fork server...

[-] PROGRAM ABORT : Timeout while initializing fork server (adjusting -t may help)
         Location : init_forkserver(), pt-fuzz-fast.c:2192
```

**Comment by junxzm1990** [comment]
Working on it. Thanks for sharing.

**Comment by evanmak** [comment]
@vanhauser-thc 
1. The binutils in the test_progs folder is a non-modified version, you can download a newer version 2.32 and try again. I guess it is something with the Docker environment. I tried building it on my test machine with native ubuntu 14.04 installed and couldn't repro the error.

2. My apology, we should've mentioned that because we modified the dynamic loader, the patched binary will not run normally unless for fuzzing with PTrix, the error output you saw is expected. We have a todo item to make a shadow copy of the target binary to patch and fuzz that instead, i think that's an appropriate fix.

3. Based on the output I can't see any reason, can you try 
1) pass `-t500+`, 
2) fuzz with another program 
3) test on a native environment (we never test it on docker before)
see if it is reproducible?

**Comment by mudongliang** [comment]
@vanhauser-thc For the 1st issue, I think you encounter [one building issue](https://sourceware.org/bugzilla/show_bug.cgi?id=22618) of binutils. Please follow the instruction - `make clean` and rebuild binutils again. If you still have problem, please remove the build directory and redo building instructions again.

**Comment by mudongliang** [comment]
@vanhauser-thc For the 2nd issue, I think the problem is that you build libtiff with `--enable-shared`(by default). Thus you encounter those problems when you try to fuzz the utilities in libtiff.
And the solution is already shown in [AFL README](https://github.com/mirrorer/afl/blob/master/docs/README#L114),

> When testing libraries, you need to find or write a simple program that reads
> data from stdin or from a file and passes it to the tested library. In such a
> case, it is essential to link this executable against a static version of the
> instrumented library, or to make sure that the correct .so file is loaded at
> runtime (usually by setting LD_LIBRARY_PATH). The simplest option is a static
> build, usually possible via:
> $ CC=/path/to/afl/afl-gcc ./configure --disable-shared

**Comment by vanhauser-thc** [comment]
@mudongliang  - yes thats fixed it thanks

@evanmak I tried many things to fuzz with pt-fuzz-fast, when I try to run from the docker container (the lkm loaded on the host) I always get the timeout. if I use your example binary or two of my own. Then I tried to run it natively on the host. I can compile basically everything (glibc does not compile completely, but it reaches building ld.so). When I fuzz that way I get

```
*] Spinning up the fork server...

[-] Hmm, looks like the target binary terminated before we could complete a
    handshake with the injected code. There are two probable explanations:
[...]
-] PROGRAM ABORT : Fork server handshake failed
         Location : init_forkserver(), pt-fuzz-fast.c:2316
```

So it seems that a) it doesnt work from inside a docker container, maybe the way the afl-pt binary communicates to the lkm doesnt work from host to container (I run the container with --security-opt seccomp:unconfined so I hoped that would allow that). and b) it does not seem to work with a current Linux (4.19.0-kali5-amd64).

damn.
